File: Python/communicator.py
import threading
from PyQt4 import QtCore
from threading import Thread
from PyQt4.QtCore import QSettings, QObject
from cryptography import generateDHPrivate, calculateDHPublic, calculateDHAES
from dialogboxes import askForInput, askForConfirmation
import dialogboxes
from exception import FileTransferBaseException
from identification import getStoredName, getStoredGuid, DEFAULT_NAME
from jsonhandling import *
import json
from multicast import UDPDiscovery
from networking import TCPSocket
from storage import *
from transferclasses import Peer, Request, PairingRequest, PublicKey, File, PreSendRequest
from strings import *

storedPeers = QSettings("FileTransfer", "peers")


class Communicator(QObject):
    requestReceivedSignal = QtCore.pyqtSignal(TCPSocket, bytes)
    peersUpdated = QtCore.pyqtSignal()
    fileReceived = QtCore.pyqtSignal(str, bytes)
    showMessageBox = QtCore.pyqtSignal(str)
    continueSend = QtCore.pyqtSignal(str, bytes, str)

    def __init__(self, parent):
        QObject.__init__(self)
        self.name = getStoredName() or DEFAULT_NAME
        self.guid = getStoredGuid()
        self.peers = self.getStoredPeers()
        self.preSendRegistrations = {}
        self.handlerMap = {}
        self.initHandlerMap()
        self.startRequestListener()
        self.requestReceivedSignal.connect(self.requestReceived)
        self.discoveryManger = UDPDiscovery()
        self.discoveryManger.peerDiscovered.connect(self.peerDiscovered)
        self.discoveryManger.peerDiscovered.connect(self.peerDiscovered)
        self.showMessageBox.connect(self.showMessageBoxSlot)
        self.discoveryManger.startDiscoveryServer()
        self.parent = parent
        self.continueSend.connect(self.continueSendSlot)

    def initHandlerMap(self):
        self.handlerMap[REQUEST_TYPE_PAIR] = self.handlePairRequest
        self.handlerMap[REQUEST_TYPE_PRE_SEND] = self.handlePreSendRequest
        self.handlerMap[REQUEST_TYPE_SEND] = self.receiveFile

    def continueSendSlot(self, fileName, fileContents, guid):
        self.sendFile(fileName, fileContents, guid)

    def showMessageBoxSlot(self, message):
        self.progressStop()
        dialogboxes.showMessageBox(self.parent, message)

    def progressRepport(self, value):
        self.parent.progressSet.emit(value)

    def progressIndeterminate(self):
        self.parent.progressIndeterminate.emit()

    def progressStop(self):
        self.parent.progressStop.emit()

    def updateName(self, newName):
        self.name = newName or DEFAULT_NAME

    def sendRequestToPeer(self, peer, type, data, timeout=None):
        peerSocket = TCPSocket()
        peerSocket.connect(self.peers[peer].lastKnownIP, 32102)
        self.sendRequestToSocket(peerSocket, type, data, timeout)
        return peerSocket

    def sendRequestToSocket(self, socket, type, data, timeout=None):
        request = objectToJSON(Request(type, data, self.guid))
        socket.setTimeout(timeout)
        socket.send(request, progressCallback=self.progressRepport)

    def peerDiscovered(self, guid, name, ip):
        if guid == self.guid:
            return
        if guid in self.peers:
            self.peers[guid].name = name
            self.peers[guid].lastKnownIP = ip
        else:
            self.peers[guid] = Peer(guid, name)
        self.storePeers()
        self.peersUpdated.emit()

    def listenerThread(self):
        listener = TCPSocket()
        listener.listen("0.0.0.0", 32102)
        while True:
            requestSocket = listener.accept()
            requestHandlerThread = Thread(target=self.receiverThread, args=(requestSocket,))
            requestHandlerThread.start()

    def receiverThread(self, socket):
        try:
            receivedData = socket.recv(self.progressRepport)
            self.progressStop()
            self.requestReceivedSignal.emit(socket, receivedData)
        except FileTransferBaseException as e:
            self.showMessageBox.emit(e.message)

    def requestReceived(self, socket, data):
        try:
            receivedRequest = JSONToObject(Request, data.decode("utf-8"))
            self.handlerMap[receivedRequest.type](socket, receivedRequest)
        except FileTransferBaseException as e:
            self.showMessageBox.emit(e.message)
        except Exception as e:
            self.showMessageBox.emit(str(e))

    def startRequestListener(self):
        thread = Thread(target=self.listenerThread, daemon=True)
        thread.start()

    def storePeers(self):
        storedPeersDict = {}
        for peer in self.peers:
            storedPeersDict[peer] = objectToJSON(self.peers[peer])
        storeItem(storedPeers, "peers", json.dumps(storedPeersDict))

    def getStoredPeers(self):
        if getStoredItem(storedPeers, "peers") is None:
            return {}
        else:
            storedPeersDict = json.loads(getStoredItem(storedPeers, "peers"))
            returnedDict = {}
            for peer in storedPeersDict:
                returnedDict[peer] = JSONToObject(Peer, storedPeersDict[peer])
            return returnedDict

    def discoverPeers(self):
        self.discoveryManger.discoverPeers()

    def pair(self, guid, continueToSend=None):
        sharedPassword = askForInput(self.parent, "Enter shared password")

        def pairThread():
            try:
                self.progressIndeterminate()
                myPrivateKey = generateDHPrivate()
                myPublicKey = calculateDHPublic(myPrivateKey)
                nonlocal sharedPassword
                if sharedPassword is None:
                    self.progressStop()
                    return False
                sharedPassword = sharedPassword
                requestPublicKey = PublicKey(myPublicKey, sharedPassword)
                pairingRequest = PairingRequest(self.name, self.guid, requestPublicKey)
                peerSocket = self.sendRequestToPeer(guid, REQUEST_TYPE_PAIR, objectToJSON(pairingRequest))
                response = peerSocket.recv().decode("utf-8")
                if response == RESPONSE_REJECT or response == RESPONSE_BAD_SIGNATURE:
                    self.showMessageBox.emit(response)
                    return False
                peersKey = JSONToObject(PublicKey, JSONToObject(Request, response).data)
                if not peersKey.verifySignature(sharedPassword):
                    self.showMessageBox.emit("Bad shared password")
                    peerSocket.send(RESPONSE_BAD_SIGNATURE)
                    return False
                peerSocket.setTimeout(5)
                peerSocket.send(RESPONSE_OK)
                self.peers[guid].myPrivateKey = myPrivateKey
                self.peers[guid].publicKey = peersKey.key
                self.peers[guid].sharedPassword = sharedPassword
                self.storePeers()
                self.peersUpdated.emit()
                self.progressStop()
                if continueToSend:
                    self.continueSend.emit(*continueToSend)
            except FileTransferBaseException as e:
                self.showMessageBox.emit(e.message)

        pairerThread = Thread(target=pairThread)
        pairerThread.start()

    def unpair(self, guid):
        if guid in self.peers:
            self.peers[guid].publicKey = self.peers[guid].myPrivateKey = self.peers[guid].sharedPassword = None
            self.storePeers()
            self.peersUpdated.emit()

    def deletePeer(self, guid):
        if guid in self.peers:
            del self.peers[guid]
            self.peersUpdated.emit()
            self.storePeers()

    def sendFile(self, fileName, fileContents, guid):
        try:
            self.progressIndeterminate()
            selectedPeer = self.peers[guid]
            filePrivateKey = generateDHPrivate()
            filePublicKey = calculateDHPublic(filePrivateKey)
            preSendPublicKey = PublicKey(filePublicKey, self.peers[guid].sharedPassword)
            preSendRequest = PreSendRequest(fileName, preSendPublicKey)
            preSendSocket = self.sendRequestToPeer(guid, REQUEST_TYPE_PRE_SEND, objectToJSON(preSendRequest))
            response = preSendSocket.recv().decode("utf-8")
            if response == RESPONSE_NOT_PAIRED:
                self.pair(guid, continueToSend=(fileName, fileContents, guid))
                return
            elif response == RESPONSE_REJECT:
                self.showMessageBox.emit("File rejected")
                self.progressStop()
                return
        except FileTransferBaseException as e:
            self.showMessageBox.emit(e.message)
            return

        def senderThread():
            try:
                aesKey = calculateDHAES(selectedPeer.publicKey, filePrivateKey)
                sentFile = File(fileName, fileContents, aesKey)
                self.sendRequestToPeer(guid, REQUEST_TYPE_SEND, objectToJSON(sentFile), 5)
                self.progressStop()
            except FileTransferBaseException as e:
                self.showMessageBox.emit(e.message)

        sThread = threading.Thread(target=senderThread)
        sThread.start()

    def handlePairRequest(self, socket, request):
        requestData = JSONToObject(PairingRequest, request.data)
        signatureSecret = askForInput(self.parent,
                                      "Pairing request from " + requestData.name + ". Enter shared password:")

        def handlerThread():
            try:
                if not signatureSecret:
                    socket.send(RESPONSE_REJECT)
                    return
                if not requestData.publicKey.verifySignature(signatureSecret):
                    self.showMessageBox.emit("Bad shared password")
                    socket.send(RESPONSE_BAD_SIGNATURE)
                    return
                myPrivateKey = generateDHPrivate()
                myPublicKey = calculateDHPublic(myPrivateKey)
                publicKey = PublicKey(myPublicKey, signatureSecret)
                self.sendRequestToSocket(socket, REQUEST_TYPE_PUBLICKEY, objectToJSON(publicKey), 5)
                response = socket.recv().decode("utf-8")
                if response != RESPONSE_OK:
                    self.showMessageBox.emit(response)
                    return
                if request.guid not in self.peers:
                    self.peers[request.guid] = Peer(request.guid, requestData.name)
                self.peers[request.guid].myPrivateKey = myPrivateKey
                self.peers[request.guid].publicKey = requestData.publicKey.key
                self.peers[request.guid].sharedPassword = signatureSecret
                self.peers[request.guid].lastKnownIP = socket.address
                self.storePeers()
                self.peersUpdated.emit()
            except FileTransferBaseException as e:
                self.showMessageBox.emit(e.message)

        hThread = Thread(target=handlerThread)
        hThread.start()

    def handlePreSendRequest(self, socket, request):
        if (request.guid in self.peers) and self.peers[request.guid].sharedPassword:
            preSendRequest = JSONToObject(PreSendRequest, request.data)
            if not preSendRequest.publicKey.verifySignature(self.peers[request.guid].sharedPassword):
                socket.send(RESPONSE_BAD_SIGNATURE)
                return
            if askForConfirmation(self.parent,
                                  "Accept file " + preSendRequest.fileName + " from " + self.peers[
                                      request.guid].name + "?"):
                self.preSendRegistrations[request.guid] = preSendRequest
                socket.send(RESPONSE_OK)
            else:
                socket.send(RESPONSE_REJECT)
        else:
            socket.send(RESPONSE_NOT_PAIRED)

    def receiveFile(self, socket, request):
        def fileReceiverThread():
            self.progressIndeterminate()
            socket.setTimeout(5)
            if not ((request.guid in self.preSendRegistrations) and self.preSendRegistrations[request.guid]):
                self.showMessageBox.emit("No presend registration")
                return
            registration = self.preSendRegistrations[request.guid]
            self.preSendRegistrations[request.guid] = False
            receivedFile = JSONToObject(File, request.data)
            if registration.fileName != receivedFile.fileName:
                self.showMessageBox.emit("Name mismatch, rejected")
                self.progressStop()
                return
            aesKey = calculateDHAES(registration.publicKey.key, self.peers[request.guid].myPrivateKey)
            fileContents = receivedFile.getContents(aesKey)
            self.fileReceived.emit(receivedFile.fileName, fileContents)
            self.progressStop()

        rThread = threading.Thread(target=fileReceiverThread)
        rThread.start()

File: WinRT/FileTransfer/FileTransfer.Shared/Communicator.cs
using FileTransfer.Cryptography;
using FileTransfer.Networking;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Threading.Tasks;
using Windows.UI.Core;

namespace FileTransfer
{
    public static class Communicator
    {
        public static ObservableCollection<Peer> Peers;
        private static Dictionary<string, Peer> peerMap;
        private static Dictionary<string, PreSendRequest> preSendRegistrations;
        private static Dictionary<string, PreSendRequestNoCrypt> preSendRegistrationsNoCrypt;
        private static string myName;
        private static TCPSocket listenerSocket;
        private static Dictionary<string, Func<string, string, TCPSocket, Task>> requestHandlerMap;
        private static CoreDispatcher uiThreadDispatcher;
        public delegate void FileReceivedHandler(string fileName, byte[] fileContents);
        public static event FileReceivedHandler FileReceived;
        public delegate void ProgressRepportHandler(int x);
        public static event ProgressRepportHandler ProgressRepport;
        public delegate void ProgressEventHandler();
        public static event ProgressEventHandler ProgressIndeterminate, ProgressStop;
        public async static Task Init(string name)
        {
            if (uiThreadDispatcher == null) uiThreadDispatcher = Windows.UI.Core.CoreWindow.GetForCurrentThread().Dispatcher;
            myName = name;
            Peers = new ObservableCollection<Peer>();
            preSendRegistrations = new Dictionary<string, PreSendRequest>();
            preSendRegistrationsNoCrypt = new Dictionary<string, PreSendRequestNoCrypt>();
            listenerSocket = new TCPSocket();
            listenerSocket.ConnectionReceived += requestReceived;
            initializeHandlerMap();
            await listenerSocket.StartListening("32102");
            if(!(await DataStorage.GetStoredItemAsync("Peers")==null))
            {
                Peer[] newPeers = JSONHandling.ParseJSONResponse<Peer[]>(await DataStorage.GetStoredItemAsync("Peers"));
                for(int i=0;i<newPeers.Length;i++) Peers.Add(newPeers[i]);
            }
            RefreshPeerMap();
            await MulticastDiscovery.StartMulticastResponder(myName);
            MulticastDiscovery.DeviceDiscovered += newPeerDiscovered;
        }
        public async static Task Reboot()
        {
            if (listenerSocket != null) listenerSocket.Close();
            listenerSocket = null;
            MulticastDiscovery.Stop();
            await Init(myName);
        }
        private static void progressStop()
        {
            if (ProgressStop != null) ProgressStop();
        }
        private static void progressIndeterminate()
        {
            if (ProgressIndeterminate != null) ProgressIndeterminate();
        }
        private static void progressRepport(int percentage)
        {
            if (ProgressRepport != null) ProgressRepport(percentage);
        }
        private async static void requestReceived(TCPSocket sender)
        {
            await uiThreadDispatcher.RunAsync(CoreDispatcherPriority.Normal, async () =>
            {
                try
                {
                    progressIndeterminate();
                    sender.ProgressRepport += (x) => { if (ProgressRepport != null) ProgressRepport(x); };
                    string requestContents = await sender.Recv();
                    progressStop();
                    Request receivedRequest = JSONHandling.ParseJSONResponse<Request>(requestContents);
                    await requestHandlerMap[receivedRequest.Type](receivedRequest.SenderGuid, receivedRequest.Data, sender);
                }
                catch(OutOfMemoryException)
                {
                    DialogBoxes.ShowMessageBox("File too big to fit in memory");
                }
                catch(FileTransferException error)
                {
                    DialogBoxes.ShowMessageBox(error.Message);
                }
                catch(Exception error)
                {
                    DialogBoxes.ShowMessageBox("Unknown error " + error.Message);
                }
                finally
                {
                    progressStop();
                    sender.Send(Strings.RESPONSE_REJECT);
                }
            });
        }

        private async static Task<TCPSocket> sendRequestToPeer(string peer, string type, string data, bool repportProgress=false, int timeout=5000)
        {
            var newSocket = new TCPSocket();
            await newSocket.Connect(peerMap[peer].LastKnownIP, "32102");
            if (repportProgress) newSocket.ProgressRepport += progressRepport;
            return await sendRequestToSocket(newSocket, type, data, timeout);
        }
        private async static Task<TCPSocket> sendRequestToSocket(TCPSocket socket, string type, string data, int timeout=5000)
        {
            var request = new Request();
            request.Data = data;
            request.SenderGuid = await GuidHandling.GetMyGuid();
            request.Type = type;
            socket.SetTimeout(timeout);
            await socket.Send(JSONHandling.SerializeObject(request));
            return socket;
        }
        public async static Task DiscoverDevices()
        {
            await MulticastDiscovery.DiscoverDevices();
        }
        public static void SetNewName(string newName) { myName = newName; MulticastDiscovery.UpdateName(myName); }
        private static void RefreshPeerMap()
        {
            peerMap = new Dictionary<string, Peer>();
            foreach(var peer in Peers)
            {
                peerMap[peer.Guid] = peer;
            }
        }
        private async static void newPeerDiscovered(string ip, string guid, string name)
        {
            await uiThreadDispatcher.RunAsync(CoreDispatcherPriority.Normal, async () =>
            {
                var newPeer = new Peer(guid, name);
                newPeer.LastKnownIP = ip;
                addPeer(newPeer);
                RefreshPeerMap();
                await StorePeers();
            });
        }
        public static Peer GetPeer(string guid)
        {
            return peerMap[guid];
        }
        public async static Task StorePeers()
        {
            Peer[] storedPeers = new Peer[Peers.Count];
            for (int i = 0; i < storedPeers.Length; i++) storedPeers[i] = Peers[i];
            await DataStorage.StoreItemAsync("Peers", JSONHandling.SerializeObject(storedPeers));
        }
        private static void initializeHandlerMap()
        {
            requestHandlerMap = new Dictionary<string, Func<string, string, TCPSocket, Task>>();
            requestHandlerMap[Strings.REQUEST_TYPE_PAIR] = pairRequestReceived;
            requestHandlerMap[Strings.REQUEST_TYPE_PRE_SEND] = preSendRequestReceived;
            requestHandlerMap[Strings.REQUEST_TYPE_SEND] = fileReceived;
            requestHandlerMap[Strings.REQUEST_TYPE_PRE_SEND_NOCRYPT] = preSendNoCryptRequestReceived;
            requestHandlerMap[Strings.REQUEST_TYPE_SEND_NOCRYPT] = fileReceivedNoCrypt;
        }
        private async static Task pairRequestReceived(string peer, string data, TCPSocket socket)
        {
            progressIndeterminate();
            var request = JSONHandling.ParseJSONResponse<PairingRequest>(data);
            if(request.Guid!=peer) 
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_REJECT);
                return;
            }
            string signatureSecret = await DialogBoxes.AskForInput("Pairing request", "Please enter the shared password");
            if (signatureSecret == null)
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_REJECT);
                return;
            }

            if(!request.PublicKey.VerifySignature(signatureSecret))
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_BAD_SIGNATURE);
                await DialogBoxes.ShowMessageBox("Shared password incorrect");
                return;
            }
            string myPrivateKey = DiffieHellman.generate_DH_Private();
            string myPublicKey = DiffieHellman.calculate_DH_Public(myPrivateKey);
            PublicKey returnedKey = new PublicKey(myPublicKey, signatureSecret);

            await sendRequestToSocket(socket, Strings.REQYEST_TYPE_PUBLICKEY, JSONHandling.SerializeObject(returnedKey));
            string response = await socket.Recv();
            progressStop();
            if(response!=Strings.RESPONSE_OK)
            {
                await DialogBoxes.ShowMessageBox(response);
                return;
            }

            Peer newPeer = new Peer(request.Guid, request.Name);
            newPeer.LastKnownIP = socket.GetIP();
            newPeer.MyPrivateKey = myPrivateKey;
            newPeer.PublicKey = request.PublicKey.Key;
            newPeer.SharedPassword = signatureSecret;
            addPeer(newPeer);
            await StorePeers();
            RefreshPeerMap();
        }
        public async static Task<bool> Pair(string guid)
        {
            try
            {
                progressIndeterminate();
                Peer pairedPeer = peerMap[guid];
                PairingRequest request = new PairingRequest();
                request.Guid = await GuidHandling.GetMyGuid();
                request.Name = myName;
                string myPrivateKey = DiffieHellman.generate_DH_Private();
                string myPublicKey = DiffieHellman.calculate_DH_Public(myPrivateKey);
                string signatureSecret = await DialogBoxes.AskForInput("Pairing", "Please enter shared password");
                if (signatureSecret == null) return false;
                request.PublicKey = new PublicKey(myPublicKey, signatureSecret);
                var socket = await sendRequestToPeer(guid, Strings.REQUEST_TYPE_PAIR, JSONHandling.SerializeObject(request), timeout:0);
                string response = await socket.Recv();
                if (response == Strings.RESPONSE_REJECT || response == Strings.RESPONSE_BAD_SIGNATURE)
                {
                    progressStop();
                    await DialogBoxes.ShowMessageBox(response);
                    return false;
                }
                PublicKey peersKey = JSONHandling.ParseJSONResponse<PublicKey>(JSONHandling.ParseJSONResponse<Request>(response).Data);
                progressStop();
                socket.SetTimeout(5000);
                if (!peersKey.VerifySignature(signatureSecret))
                {
                    await DialogBoxes.ShowMessageBox("Bad shared password");
                    await socket.Send(Strings.RESPONSE_BAD_SIGNATURE);
                    return false;
                }
                await socket.Send(Strings.RESPONSE_OK);
                Peer newPeer = new Peer(guid, pairedPeer.Name);
                newPeer.MyPrivateKey = myPrivateKey;
                newPeer.PublicKey = peersKey.Key;
                newPeer.LastKnownIP = socket.GetIP();
                newPeer.SharedPassword = signatureSecret;
                addPeer(newPeer);
                await StorePeers();
                RefreshPeerMap();
                return true;
            }
            catch(FileTransferException error)
            {
                DialogBoxes.ShowMessageBox(error.Message);
                return false;
            }
            catch(Exception error)
            {
                DialogBoxes.ShowMessageBox("Unknown error: " + error.Message);
                return false;
            }
            finally
            {
                progressStop();
            }
        }

        public async static Task Unpair(string guid)
        {
            Peer selectedPeer = peerMap[guid];
            Peer newPeer = new Peer(selectedPeer.guid, selectedPeer.Name);
            newPeer.LastKnownIP = selectedPeer.LastKnownIP;
            Peers.Remove(selectedPeer);
            Peers.Add(newPeer);
            RefreshPeerMap();
            await StorePeers();
        }
        private static void addPeer(Peer addedPeer)
        {
            if(peerMap.ContainsKey(addedPeer.Guid))
            {
                var oldPeer = peerMap[addedPeer.Guid];
                Peers.Remove(oldPeer);
                if (oldPeer.PublicKey != null && addedPeer.PublicKey == null)
                {
                    addedPeer.PublicKey = oldPeer.PublicKey;
                    addedPeer.MyPrivateKey = oldPeer.MyPrivateKey;
                }
                if (oldPeer.SharedPassword != null && addedPeer.SharedPassword == null)
                {
                    addedPeer.SharedPassword = oldPeer.SharedPassword;
                }
            }
            Peers.Add(addedPeer);
            RefreshPeerMap();
        }
        public async static Task SendFile(string fileName, byte[] fileContents, string peer)
        {
            try
            {
                progressIndeterminate();
                Peer selectedPeer = peerMap[peer];
                PreSendRequest preSend = new PreSendRequest();
                String filePrivateKey = DiffieHellman.generate_DH_Private();
                String filePublicKey = DiffieHellman.calculate_DH_Public(filePrivateKey);
                byte[] aesKey = DiffieHellman.calculate_DH_AES(selectedPeer.PublicKey, filePrivateKey);
                preSend.FileName = fileName;
                preSend.key = new PublicKey(filePublicKey, selectedPeer.SharedPassword);
                var preSendSocket = await sendRequestToPeer(peer, Strings.REQUEST_TYPE_PRE_SEND, JSONHandling.SerializeObject(preSend), timeout:0);
                var response = await preSendSocket.Recv();
                if (response == Strings.RESPONSE_NOT_PAIRED)
                {
                    if (!(await Pair(peer))) return;
                    await SendFile(fileName, fileContents, peer);
                    return;
                }
                else if (response == Strings.RESPONSE_REJECT)
                {
                    await DialogBoxes.ShowMessageBox("File rejected");
                    progressStop();
                    return;
                }
                File sentFile = new File(fileName, fileContents, aesKey);
                await sendRequestToPeer(peer, Strings.REQUEST_TYPE_SEND, JSONHandling.SerializeObject(sentFile), true);
                progressStop();
            }
            catch (FileTransferException error)
            {
                DialogBoxes.ShowMessageBox(error.Message);
            }
            catch (Exception error)
            {
                DialogBoxes.ShowMessageBox("Unknown error: " + error.Message);
            }
            finally
            {
                progressStop();
            }
        }
        public async static Task SendFileNoCrypt(string fileName, byte[] fileContents, string peer)
        {
            try
            {
                progressIndeterminate();
                PreSendRequestNoCrypt preSend = new PreSendRequestNoCrypt();
                preSend.FileName = fileName;
                var preSendSocket = await sendRequestToPeer(peer, Strings.REQUEST_TYPE_PRE_SEND_NOCRYPT, JSONHandling.SerializeObject(preSend), timeout: 0);
                var response = await preSendSocket.Recv();
                if (response == Strings.RESPONSE_NOT_PAIRED)
                {
                    if (!(await Pair(peer))) return;
                    await SendFile(fileName, fileContents, peer);
                    return;
                }
                else if (response == Strings.RESPONSE_REJECT)
                {
                    await DialogBoxes.ShowMessageBox("File rejected");
                    progressStop();
                    return;
                }
                File sentFile = new File(fileName, fileContents);
                await sendRequestToPeer(peer, Strings.REQUEST_TYPE_SEND_NOCRYPT, JSONHandling.SerializeObject(sentFile), true);
                progressStop();
            }
            catch (FileTransferException error)
            {
                DialogBoxes.ShowMessageBox(error.Message);
            }
            catch (Exception error)
            {
                DialogBoxes.ShowMessageBox("Unknown error: " + error.Message);
            }
            finally
            {
                progressStop();
            }
        }
        private async static Task fileReceived(string peer, string data, TCPSocket socket)
        {
            progressIndeterminate();
            if(!(peerMap.ContainsKey(peer) && peerMap[peer].Paired))
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_NOT_PAIRED);
                return;
            }
            if (!(preSendRegistrations.ContainsKey(peer)))
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_REJECT);
                return;
            }
            PreSendRequest registration = preSendRegistrations[peer];
            preSendRegistrations.Remove(peer);
            File receivedFile = JSONHandling.ParseJSONResponse<File>(data);
            if(receivedFile.FileName!=registration.FileName)
            {
                progressStop();
                throw new FileTransferException("File name mismatch");
            }
            byte[] aesKey = DiffieHellman.calculate_DH_AES(registration.key.Key, peerMap[peer].MyPrivateKey);
            byte[] fileContents = receivedFile.GetContents(aesKey);
            progressStop();
            FileReceived(receivedFile.FileName, fileContents);
        }

        private async static Task fileReceivedNoCrypt(string peer, string data, TCPSocket socket)
        {
            progressIndeterminate();
            if (!(peerMap.ContainsKey(peer) && peerMap[peer].Paired))
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_NOT_PAIRED);
                return;
            }
            if (!(preSendRegistrationsNoCrypt.ContainsKey(peer)))
            {
                progressStop();
                await socket.Send(Strings.RESPONSE_REJECT);
                return;
            }
            PreSendRequestNoCrypt registration = preSendRegistrationsNoCrypt[peer];
            preSendRegistrationsNoCrypt.Remove(peer);
            File receivedFile = JSONHandling.ParseJSONResponse<File>(data);
            if (receivedFile.FileName != registration.FileName)
            {
                progressStop();
                throw new FileTransferException("File name mismatch");
            }
            byte[] fileContents = receivedFile.GetContents();
            progressStop();
            FileReceived(receivedFile.FileName, fileContents);
        }

        private async static Task preSendRequestReceived(string peer, string data, TCPSocket socket)
        {
            if(peerMap.ContainsKey(peer) && peerMap[peer].Paired)
            {
                PreSendRequest registration = JSONHandling.ParseJSONResponse<PreSendRequest>(data);
                if(!registration.key.VerifySignature(peerMap[peer].SharedPassword))
                {
                    socket.Send(Strings.RESPONSE_REJECT);
                }
                await DialogBoxes.AskForConfirmation(registration.FileName+" from "+peerMap[peer].Name+". Accept?", async () =>
                {
                    preSendRegistrations[peer] = registration;
                    await socket.Send(Strings.RESPONSE_OK);
                }, async () =>
                {
                    await socket.Send(Strings.RESPONSE_REJECT);
                });
            }
            else
            {
                await socket.Send(Strings.RESPONSE_NOT_PAIRED);
            }
        }

        private async static Task preSendNoCryptRequestReceived(string peer, string data, TCPSocket socket)
        {
            if (peerMap.ContainsKey(peer) && peerMap[peer].Paired)
            {
                PreSendRequestNoCrypt registration = JSONHandling.ParseJSONResponse<PreSendRequestNoCrypt>(data);
                await DialogBoxes.AskForConfirmation(registration.FileName + " from " + peerMap[peer].Name + " over insecure connection. Accept?", async () =>
                {
                    preSendRegistrationsNoCrypt[peer] = registration;
                    await socket.Send(Strings.RESPONSE_OK);
                }, async () =>
                {
                    await socket.Send(Strings.RESPONSE_REJECT);
                });
            }
            else
            {
                await socket.Send(Strings.RESPONSE_NOT_PAIRED);
            }
        }

        public async static Task DeletePeer(string selectedGuid)
        {
            Peers.Remove(peerMap[selectedGuid]);
            peerMap.Remove(selectedGuid);
            await StorePeers();
        }
    }
}
